addons/Hierarchical_Lesson_Report/src/presenter.js

```diff
@@ -72,135 +123,308 @@ function AddonHierarchical_Lesson_Report_create() {
+        if (configuration.showMaxScoreField)
+            addMaxScoreAwardCell($row, labels.maxScoreAward, isInPrintableMode);
+
+        $row.prependTo($view);
     }
 
-    presenter.calculateLessonScaledScore = function () {
-        if (presenter.lessonScore.pageCount == 0) {
-            return 0;
-        }
+    function generateHeaderRow(isInPrintableMode) {
+        var $row = $(document.createElement('tr'));
+        var cssClass = isInPrintableMode
+            ? CSS_CLASSES.PRINTABLE_HIER_REPORT_HEADER
+            : CSS_CLASSES.HIER_REPORT_HEADER;
+        $row.addClass(cssClass);
+        return $row;
+    }
 
-        return Math.floor((presenter.lessonScore.scaledScore / presenter.lessonScore.pageCount) * 100) / 100;
-    };
+    function addTitleHeaderCell($row, title, isInPrintableMode) {
+        var $cell =  $(document.createElement('td'));
+        if (isInPrintableMode)
+            $cell.addClass(CSS_CLASSES.PRINTABLE_HIER_REPORT_TITLE_HEADER);
+        $cell.html(title);
+        $cell.appendTo($row);
+    }
 
-    function addFooter() {
-        var row = document.createElement('tr');
-        $(row).appendTo($("#" + presenter.treeID).find('table'));
-        $(row).addClass("hier_report-footer");
+    function addResultsHeaderCell($row, header, isInPrintableMode) {
+        var $cell = $(document.createElement('td'));
+        var cssClass = isInPrintableMode
+            ? CSS_CLASSES.PRINTABLE_HIER_REPORT_PROGRESS_HEADER
+            : CSS_CLASSES.HIER_REPORT_PROGRESS;
+        $cell.addClass(cssClass);
+        $cell.html(header);
+        $cell.appendTo($row);
+    }
 
-        $("<td></td>").appendTo($(row)).html(presenter.configuration.labels.total);
+    function addChecksCell($row, checks, isInPrintableMode) {
+        var $cell = $(document.createElement('td'));
+        var cssClass = isInPrintableMode
+            ? CSS_CLASSES.PRINTABLE_HIER_REPORT_CHECKS
+            : CSS_CLASSES.HIER_REPORT_CHECKS;
+        $cell.addClass(cssClass);
+        $cell.html(checks);
+        $cell.appendTo($row);
+    }
 
-        if (presenter.configuration.showResults) {
-            createProgressCell(row, {
-                score: presenter.isPreview ? 0 : presenter.calculateLessonScaledScore(),
-                count: 1
-            });
-        }
+    function addMistakesCell($row, mistakes, isInPrintableMode) {
+        var $cell = $(document.createElement('td'));
+        var cssClass = isInPrintableMode
+            ? CSS_CLASSES.PRINTABLE_HIER_REPORT_MISTAKES
+            : CSS_CLASSES.HIER_REPORT_MISTAKES;
+        $cell.addClass(cssClass);
+        $cell.html(mistakes);
+        $cell.appendTo($row);
+    }
 
-        if (presenter.configuration.showChecks) {
-            $("<td></td>").appendTo($(row)).addClass("hier_report-checks").html(presenter.lessonScore.checks);
-        }
+    function addErrorsCell($row, errors, isInPrintableMode) {
+        var $cell = $(document.createElement('td'));
```

Comment:
Te funkcje są bardzo podobne - tworzą htmlową komórkę, dodają klasę, wrzucają kontent i dodają do komórki. Jeżeli stworzone by zostały dwie CSS_CLASSES - jedna z printable druga z normalnymi, to można by w funkcji addHeader zrobić coś w stylu:
`var CSS_CLASSES = isPrintableMode ?  PRINTABLE_CSS_CLASSES : DISPLAY_CSS_CLASSESS`
Następnie wywołanie każdej funkcji byłoby z przekazaniem argumentu gotowej klasy CSS, np addDataToCell(row, label.error, CSS_CLASSES.HIER_REPORT_ERRORS)
Są też funkcje, jak maxScoreAward, która sprawdza maxScoreAward, ale równie dobrze jeżeli parametry byłyby puste (finalnie zostanie dodany pusty string/null) to można spokojnie wykorzystać generyczną funkcję.

Reply:
Temat został omówiony na spotkaniu